phac-nml/irida-next
# GraphQL: Clone Samples and Transfer Samples mutations

## What does this PR do and why?
_Describe in detail what your merge request does and why._
Fixes #809
Fixes #810

Adds two new mutations for cloning and transfering samples

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Sign into admin and go to localhost:3000/graphiql
2. Get some info on projects you want to use
```graphql
query getProject1{
  project(fullPath: "bacillus/bacillus-anthracis/outbreak-2022") {
    id
    name
    path
    samples{
      nodes{
        name
        id
      }
    }
  }
}

query getProject2{
  project(fullPath: "bacillus/bacillus-anthracis/outbreak-2023") {
    id
    name
    path
    samples{
      nodes{
        name
        id
      }
    }
  }
}

```
3. Try transferring samples between projects. You can verify them with the project info queries above, or by navigating to them in the IRIDA Next web ui
```graphql
mutation transfer_samps{
  transferSamples(input:{
    projectId: "gid://irida/Project/e0bba0d9-84e9-490a-ba84-e0c0c913cdaf",
    newProjectId: "gid://irida/Project/d86955eb-af27-4a43-a01d-e6c93bb7dc6f",
    sampleIds:[
      "gid://irida/Sample/d79bdb22-92c9-4b98-87df-b4e9f79cab4d", 
      "gid://irida/Sample/cc1aa061-03cd-48da-96d7-a21e93613d43"
    ]
  }){
    samples,
    errors{
      path
      message
    }
  }
}
```
4. Try copying samples from one project to another. You can verify them with the project info queries above, or by navigating to them in the IRIDA Newx web ui
```graphql
mutation copy_samples{
  copySamples(input:{
    projectId: "gid://irida/Project/e0bba0d9-84e9-490a-ba84-e0c0c913cdaf",
    newProjectId: "gid://irida/Project/d86955eb-af27-4a43-a01d-e6c93bb7dc6f",
    sampleIds:[
      "gid://irida/Sample/d79bdb22-92c9-4b98-87df-b4e9f79cab4d", 
      "gid://irida/Sample/cc1aa061-03cd-48da-96d7-a21e93613d43"
    ]
  }){
    samples,
    errors{
      path
      message
    }
  }
}
```

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
